File: app/src/main/java/com/sunny/youyun/wifidirect/activity/single_trans/trans/TransModel.java
package com.sunny.youyun.wifidirect.activity.single_trans.trans;

import android.os.Handler;

import com.orhanobut.logger.Logger;
import com.sunny.youyun.utils.UUIDUtil;
import com.sunny.youyun.wifidirect.event.FileTransEvent;
import com.sunny.youyun.wifidirect.manager.SingleTransManager;
import com.sunny.youyun.wifidirect.manager.WifiDirectManager;
import com.sunny.youyun.wifidirect.model.TransLocalFile;
import com.sunny.youyun.wifidirect.server.SingleTransMode;
import com.sunny.youyun.wifidirect.utils.TransRxBus;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Created by Sunny on 2017/8/11 0011.
 */

class TransModel implements TransContract.Model {
    private TransPresenter mPresenter;
    private final String uuid = UUIDUtil.getUUID();
    private final List<TransLocalFile> mList = new CopyOnWriteArrayList<>();
    private final Handler handler = new Handler();


    TransModel(TransPresenter transPresenter) {
        mPresenter = transPresenter;
    }

    @Override
    public void begin() {
        TransRxBus.getInstance()
                .subscribe(uuid, fileTransEvent -> {
                    Logger.i(fileTransEvent.toString());
                    switch (fileTransEvent.getType()){
                        case FINISH:
                        case BEGIN:
                            handler.post(()-> mPresenter.update());
                            break;
                        case UPLOAD:
                        case DOWNLOAD:
                            System.out.println(fileTransEvent.getAlready());
                            if (fileTransEvent.isDone() || fileTransEvent.getType() == FileTransEvent.Type.BEGIN)
                                handler.post(() -> mPresenter.update());
                            if(mList.size() <= fileTransEvent.getPosition()) {
                                Logger.i("mList.size() > fileTransEvent.getPosition()");
                                return;
                            }
                            handler.post(() -> mPresenter.updateItem(mList.size() - fileTransEvent.getPosition() - 1, mList.get(fileTransEvent.getPosition())));
                            break;
                        case ERROR:
                            mPresenter.showError("传输失败");
                            break;
                    }
                }, throwable -> Logger.e(throwable, "面对面快传进度监听错误"));
        SingleTransMode.getInstance()
                .bindMList(mList);
        WifiDirectManager.getInstance().startServer(SingleTransMode.getInstance());
    }

    @Override
    public void exit() {
        TransRxBus.getInstance().unSubscribe(uuid);
    }

    @Override
    public void send(String[] paths) {
        System.out.println("ip: " + SingleTransManager.getInstance().getTargetInfo().getIp());
        System.out.println(SingleTransManager.getInstance().getTargetInfo());
        for (String path : paths) {
            WifiDirectManager.getInstance()
                    .sendFile(path, SingleTransManager.getInstance().getTargetInfo().getIp(), mList);
        }
    }

    @Override
    public List<TransLocalFile> getData() {
        return mList;
    }

}

File: app/src/main/java/com/sunny/youyun/wifidirect/activity/single_trans/trans/TransModel.kt
package com.sunny.youyun.wifidirect.activity.single_trans.trans

import android.os.Handler

import com.orhanobut.logger.Logger
import com.sunny.youyun.App
import com.sunny.youyun.utils.UUIDUtil
import com.sunny.youyun.wifidirect.event.FileTransEvent
import com.sunny.youyun.wifidirect.model.TransLocalFile
import com.sunny.youyun.wifidirect.utils.TransRxBus
import com.sunny.youyun.wifidirect.socket.client.Client
import com.sunny.youyun.wifidirect.socket.server.Server
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Created by Sunny on 2017/8/11 0011.
 */

class TransModel(private val mPresenter: TransPresenter) : TransContract.Model {
    private val uuid = UUIDUtil.getUUID()
    private val mList = CopyOnWriteArrayList<TransLocalFile>()
    private val handler = Handler()

    override val data: List<TransLocalFile>
        get() = mList

    override fun begin() {
        TransRxBus.getInstance()
                .subscribe(uuid, { fileTransEvent ->
                    Logger.i(fileTransEvent.toString())
                    when (fileTransEvent.type ?: -1) {
                        FileTransEvent.Type.FINISH -> {
                            if (fileTransEvent.type == FileTransEvent.Type.UPLOAD) {
                                App.mList_SendRecord.add(mList[fileTransEvent.position])
                                return@subscribe
                            }
                            if (fileTransEvent.type == FileTransEvent.Type.DOWNLOAD) {
                                App.mList_ReceiveRecord.add(mList[fileTransEvent.position])
                            }
                            handler.post { mPresenter.update() }
                        }
                        FileTransEvent.Type.BEGIN -> handler.post { mPresenter.update() }
                        FileTransEvent.Type.UPLOAD, FileTransEvent.Type.DOWNLOAD -> {
                            println(fileTransEvent.already)
                            if (fileTransEvent.isDone || fileTransEvent.type == FileTransEvent.Type.BEGIN)
                                handler.post { mPresenter.update() }
                            if (mList.size <= fileTransEvent.position) {
                                Logger.i("mList.size() > fileTransEvent.getPosition()")
                                return@subscribe
                            }
                            handler.post { mPresenter.updateItem(mList.size - fileTransEvent.position - 1, mList[fileTransEvent.position]) }
                        }
                        FileTransEvent.Type.ERROR -> mPresenter.showError("传输失败")
                    }
                }) { throwable -> Logger.e(throwable, "面对面快传进度监听错误") }
        Server.INSTANCE
                .startSync()
        Server.INSTANCE
                .startSync()
    }

    override fun exit() {
        TransRxBus.getInstance().unSubscribe(uuid)
    }

    override fun send(ip: String, paths: Array<String>) {
        for (path in paths) {
            Client.INSTANCE
                    .sendFile(path, ip, mList)
        }
    }

}
